include/exadg/grid/grid.h

```diff
@@ -73,6 +74,124 @@ class Grid
     mapping = std::make_shared<dealii::MappingQGeneric<dim>>(data.mapping_degree);
   }
 
+  void
+  create_and_refine_triangulation(
+    GridData const &                                          data,
+    std::vector<int> const &                                  vector_local_refinements,
+    std::function<void(dealii::Triangulation<dim> &)> const & create_coarse_triangulation)
+  {
+    if(data.triangulation_type == TriangulationType::Serial or
+       data.triangulation_type == TriangulationType::Distributed)
+    {
+      create_coarse_triangulation(*triangulation);
+
+      if(vector_local_refinements.size() > 0)
+        refine_local(*triangulation, vector_local_refinements);
+
+      triangulation->refine_global(data.n_refine_global);
+    }
+    else if(data.triangulation_type == TriangulationType::FullyDistributed)
+    {
+      auto const serial_grid_generator = [&](dealii::Triangulation<dim, dim> & tria_serial) {
+        create_coarse_triangulation(tria_serial);
+
+        if(vector_local_refinements.size() > 0)
+          refine_local(tria_serial, vector_local_refinements);
+
+        tria_serial.refine_global(data.n_refine_global);
+      };
+
+      auto const serial_grid_partitioner = [&](dealii::Triangulation<dim, dim> & tria_serial,
+                                               MPI_Comm const                    comm,
+                                               unsigned int const                group_size) {
+        (void)group_size;
+        if(data.partitioning_type == PartitioningType::Metis)
+        {
+          dealii::GridTools::partition_triangulation(dealii::Utilities::MPI::n_mpi_processes(comm),
+                                                     tria_serial);
+        }
+        else if(data.partitioning_type == PartitioningType::p4est)
+        {
+          dealii::GridTools::partition_triangulation_zorder(
+            dealii::Utilities::MPI::n_mpi_processes(comm), tria_serial);
+        }
+        else
+        {
+          AssertThrow(false, dealii::ExcNotImplemented());
+        }
+      };
+
+      unsigned int const group_size = 1;
+
+      auto const description = dealii::TriangulationDescription::Utilities::
+        create_description_from_triangulation_in_groups<dim, dim>(
+          serial_grid_generator,
+          serial_grid_partitioner,
+          triangulation->get_communicator(),
+          group_size,
+          dealii::Triangulation<dim>::none,
+          dealii::TriangulationDescription::construct_multigrid_hierarchy);
```

Comment:
This will not not work with simplex/mixed meshes.

Reply:
what exactly? the whole function call or just certain parameters?

Reply:
I still think it might be good to have an input parameter that decides whether the mesh is hypercube/simplex instead of deducing that information from the triangulation.

Assume we add such an `enum class` parameter to `GridData`, e.g. `enum class ElementType {Hypercube, Simplex};`. Then, the function `create_triangulation()` could check in the end that the created triangulation is consistent with the input parameter. The rest of the code would then work with the parameter `ElementType` and not with `Triangulation` (and `all_cells_are_hypercube/simplex()`).

Reply:
@peterrum, what would I have to change? Maybe you could write this to PR #131 dealing with simplex (see comment below, where I suggest to postpone simplex-content in this PR and address it in PR #131).